kv: TestTxnCoordSenderHeartbeat is flaky

Discovered in #4750 but confirmed on master. I first suspected some version of https://github.com/golang/go/issues/14539 but then also reproduced it on go tip.

Here's what a successful run looks like
```
=== RUN   TestTxnCoordSenderHeartbeat
I160302 15:17:30.116319 storage/replica_command.go:1272  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1.000s
E160302 15:17:30.116557 config/config.go:352  unable to determine largest reserved object ID from system config: descriptor table not found in system config of 0 values
I160302 15:17:30.119632 stopper.go:273  draining; tasks left:
1      kv/txn_coord_sender.go:822
--- PASS: TestTxnCoordSenderHeartbeat (0.01s)
PASS
ok  	github.com/cockroachdb/cockroach/kv	0.034s
```

And a failure:
```
make stress PKG=./kv TESTS=TestTxnCoordSenderHeartbeat STRESSFLAGS='-stderr -maxfails 1'
16 runs so far, 0 failures, over 20s

I160302 15:10:48.626091 storage/replica_command.go:1272  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1.000s
E160302 15:10:48.626349 config/config.go:352  unable to determine largest reserved object ID from system config: descriptor table not found in system config of 0 values

15000 lines of:

W160302 15:10:48.628474 kv/txn_coord_sender.go:680  heartbeat to id=4d59dcbb key="a" rw=true pri=0.01074560 iso=SERIALIZABLE stat=PENDING epo=0 ts=0.000000000,4 orig=0.000000000,4 max=0.000000000,4 failed: retry txn id=4d59dcbb key="a" rw=true pri=0.01074560 iso=SERIALIZABLE stat=PENDING epo=0 ts=0.000000000,4 orig=0.000000000,4 max=0.000000000,4

Finally:

--- FAIL: TestTxnCoordSenderHeartbeat (20.14s)
	testing.go:170: kv/txn_coord_sender_test.go:284: condition failed to evaluate within 15s: kv/txn_coord_sender_test.go:272: got txn: false: retry txn id=4d59dcbb key="a" rw=true pri=0.01447937 iso=SERIALIZABLE stat=PENDING epo=1 ts=0.000000029,38859 orig=0.000000029,38859 max=0.000000000,4
	leaktest.go:80: Leaked goroutine: goroutine 11 [chan receive]:
		github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open.func1(0xc8201d38b0)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:123 +0x3a
		created by github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:124 +0x32a
	leaktest.go:80: Leaked goroutine: goroutine 20 [select]:
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).startStats(0xc82013a7e0)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:253 +0xae6
		github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).(github.com/cockroachdb/cockroach/kv.startStats)-fm()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:238 +0x20
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc820116bb0)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
	leaktest.go:80: Leaked goroutine: goroutine 21 [select]:
		github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/store.go:1768 +0x608
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc820128de0)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
	leaktest.go:80: Leaked goroutine: goroutine 22 [select]:
		github.com/cockroachdb/cockroach/storage.(*Store).Start.func3()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/store.go:688 +0x12b
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc8200c7200)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
	leaktest.go:80: Leaked goroutine: goroutine 23 [select]:
		github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func1()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/store.go:748 +0x2c5
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc8200c7240)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
	leaktest.go:80: Leaked goroutine: goroutine 24 [select]:
		github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func2()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/store.go:767 +0x2c5
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc8200c7260)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
	leaktest.go:80: Leaked goroutine: goroutine 50 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/queue.go:351 +0x290
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc8201d0580)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
	leaktest.go:80: Leaked goroutine: goroutine 51 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/queue.go:351 +0x290
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc8201d05e0)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
	leaktest.go:80: Leaked goroutine: goroutine 52 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/queue.go:351 +0x290
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc8201d0600)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
	leaktest.go:80: Leaked goroutine: goroutine 53 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/queue.go:351 +0x290
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc8201d0620)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
	leaktest.go:80: Leaked goroutine: goroutine 54 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/queue.go:351 +0x290
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc8201d0640)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
	leaktest.go:80: Leaked goroutine: goroutine 55 [select]:
		github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/queue.go:351 +0x290
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc8201d0660)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
	leaktest.go:80: Leaked goroutine: goroutine 56 [select]:
		github.com/cockroachdb/cockroach/storage.(*replicaScanner).waitAndProcess(0xc82013a900, 0xece693fc8, 0xc8254e0336, 0x5a3a260, 0xc820217c80, 0xc8202299d0, 0xc82006e410, 0xc820461ed8)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:176 +0x31e
		github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1.1(0xc82006e410, 0xc82014e150)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:214 +0x70
		github.com/cockroachdb/cockroach/storage.(*storeRangeSet).Visit(0xc82014e750, 0xc82015e8c0)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/store.go:229 +0x226
		github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1()
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:216 +0x1a0
		github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8202299d0, 0xc8201d0680)
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x4c
		created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
			/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x5c
FAIL


ERROR: exit status 1

17 runs completed, 1 failures, over 21s
FAIL
```

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -1355,11 +1355,12 @@ func (r *Replica) applyRaftCommandInBatch(ctx context.Context, index uint64, ori
 	btch := r.store.Engine().NewBatch()
 
 	// Check the sequence for this batch. Only applies to transactional
-	// requests: on a cache hit, the transaction is instructed to restart.
-	// The epoch check allows us to poison the sequence cache to avoid
+	// requests which write intents (for example HeartbeatTxn does not get
+	// hindered by this). On a cache hit, the transaction is instructed to
+	// restart. The epoch check allows us to poison the sequence cache to avoid
 	// anomalies when we know that the transaction has been aborted or pushed,
 	// but it itself does not.
-	if ba.IsWrite() && ba.Txn != nil {
+	if ba.Txn != nil && ba.IsTransactionWrite() {
 		if err := r.checkSequenceCache(btch, *ba.Txn); err != nil {
 			return btch, nil, nil, err
 		}
@@ -1414,15 +1415,16 @@ func (r *Replica) applyRaftCommandInBatch(ctx context.Context, index uint64, ori
 			*ms = engine.MVCCStats{}
 		}
 
-		// Only transactional requests have replay protection. If the transaction
-		// just ended (via a successful EndTransaction call), it has purged its
-		// sequence cache state and would prefer if we didn't write an entry
-		// which would then need to be GCed on the slow path. See
-		// (*Replica).adjustResponseCache for a more salient explanation of
-		// why that is ok.
 		if ba.Txn != nil {
+			// Only transactional requests which leave intents have replay
+			// protection. In the particular case of a txn which just ended
+			// (via a successful EndTransaction call), it has purged its
+			// sequence cache state and would prefer if we didn't write an
+			// entry which would then need to be GCed on the slow path. See
+			// (*Replica).adjustResponseCache for a more salient explanation
+			// of why that is ok.
 			_, isEndTxn := ba.GetArg(roachpb.EndTransaction)
-			if !isEndTxn || err != nil {
+			if (!isEndTxn && ba.IsTransactionWrite()) || err != nil {
 				if putErr := r.sequence.Put(btch, ms, ba.Txn.ID, ba.Txn.Epoch,
 					ba.Txn.Sequence, ba.Txn.Key, ba.Txn.Timestamp, err); putErr != nil {
 					// TODO(tschottdorf): ReplicaCorruptionError.
@@ -1466,7 +1468,8 @@ func (r *Replica) checkSequenceCache(b engine.Engine, txn roachpb.Transaction) *
 	if err != nil {
 		// We hit the cache, so let the transaction restart.
 		if log.V(1) {
-			log.Infof("found sequence cache entry for %s@%d", txn.Short(), txn.Sequence)
+			log.Infof("found sequence cache entry for %s@epo=%d,seq=%d: epo=%d,seq=%d", txn.Short(),
+				txn.Epoch, txn.Sequence, epoch, sequence)
 		}
 		newTxn := txn.Clone()
 		newTxn.Timestamp.Forward(entry.Timestamp)

--- a/storage/replica_test.go
+++ b/storage/replica_test.go
@@ -1617,6 +1617,29 @@ func TestRangeSequenceCacheStoredTxnRetryError(t *testing.T) {
 	}
 }
 
+// TestRangeSequenceCacheOnlyWithIntent verifies that a transactional command
+// which goes through Raft but is not a transactional write (i.e. does not
+// leave intents) passes the sequence cache unhindered.
+func TestRangeSequenceCacheOnlyWithIntent(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	tc := testContext{}
+	tc.Start(t)
+	defer tc.Stop()
+
+	txn := newTransaction("test", []byte("test"), 10, roachpb.SERIALIZABLE, tc.clock)
+	txn.Sequence = 100
+	if err := tc.rng.sequence.Put(tc.engine, nil, txn.ID, uint32(txn.Epoch), txn.Sequence, txn.Key, txn.Timestamp, nil); err != nil {
+		t.Fatal(err)
+	}
+
+	args, h := heartbeatArgs(txn)
+	// If the sequence cache were active for this request, we'd catch a txn retry.
+	// Instead, we expect the error from heartbeating a nonexistent txn.
+	if _, pErr := client.SendWrappedWith(tc.Sender(), tc.rng.context(), h, &args); !testutils.IsPError(pErr, "record not present") {
+		t.Fatal(pErr)
+	}
+}
+
 // TestEndTransactionDeadline verifies that EndTransaction respects the
 // transaction deadline.
 func TestEndTransactionDeadline(t *testing.T) {

